gcp_variant_transforms/bq_to_vcf.py

```diff
@@ -56,19 +63,137 @@ def run(argv=None):
+         | beam.ParDo(_get_file_path_and_sorted_variants, bq_to_vcf_temp_folder)
+         | vcfio.WriteVcfDataLines())
+
+
+def _compose_temp_files(project, output_file, bq_to_vcf_temp_folder):
+  # type: (str, str, str) -> None
+  """Composes intermediate files to one VCF file.
+
+  It composes VCF data shards to one VCF data file and deletes the intermediate
+  VCF shards.
+  TODO(allieychen): Eventually, it further consolidates the meta information,
+  data header line, and the composed VCF data file into the `output_file`.
+  TODO(allieychen): Move the composing logic into a separate library.
+  """
+  bucket_name, blob_prefix = gcsio.parse_gcs_path(bq_to_vcf_temp_folder)
+  client = storage.Client(project)
+  bucket = client.get_bucket(bucket_name)
+  composed_vcf_data_file = _compose_vcf_shards_to_one(bucket, blob_prefix)
+  output_file_blob = _create_blob(client, output_file)
+  output_file_blob.rewrite(composed_vcf_data_file)
+  bucket.delete_blobs(bucket.list_blobs(prefix=blob_prefix))
+
+
+def _get_file_path_and_sorted_variants((file_name, variants), file_path_prefix):
+  # type: (Tuple[str, List], str) -> Iterable[Tuple[str, List]]
+  """Returns the file path and the sorted variants.
+
+  Args:
+    file_name: The file name that associated with the `variants`, which is used
+      to form the file path where the `variants` are written to.
+    variants: A collection of variants within a contiguous region of the genome.
+    file_path_prefix: The common file prefix for all VCF files generated in the
+      pipeline. The files written will begin with this prefix, followed by the
+      `file_name`.
+  """
+  file_path = '/'.join([file_path_prefix, file_name])
+  yield (file_path, sorted(variants))
+
+
+def _pair_variant_with_key(variant, number_of_variants_per_shard):
+  return ('%s_%011d' % (variant.reference_name,
+                        variant.start / number_of_variants_per_shard *
+                        number_of_variants_per_shard),
+          variant)
+
+
+def _compose_vcf_shards_to_one(bucket, blob_prefix):
+  # type: (storage.Bucket, str) -> storage.Blob
+  """Composes multiple VCF shards in GCS to one.
+
+  Note that Cloud Storage allows to compose up to 32 objects. This method
+  composes the VCF files recursively until there is only one file.
+
+  Args:
+    bucket: the bucket in which the VCF files will be composed.
+    blob_prefix: the prefix used to filter blobs. Only the VCF files with this
+      prefix will be composed.
+
+  Returns:
+    The final blob that all VCFs composed to.
+  """
+  blobs_to_be_composed = list(bucket.list_blobs(prefix=blob_prefix))
+  if len(blobs_to_be_composed) == 1:
+    return blobs_to_be_composed[0]
+  new_blob_prefix = filesystems.FileSystems.join(blob_prefix, 'composed_')
+  for blobs_chunk in (_break_list_in_chunks(blobs_to_be_composed,
+                                            _MAX_NUM_OF_BLOBS_PER_COMPOSE)):
+    _, file_name = os.path.split(blobs_chunk[0].name)
```

Comment:
nit: does FileSystem(s) have a method for this?

Reply:
Thank you! Nice advice!